### ISSUE TITLE
Audit Trail with Prisma 2

### ISSUE THREAD

**Issue by yuval-hazaz** [prisma/prisma2#1902]
## Problem
The project is currently missing methods to customize the process. Hooks or lifecycle events are needed in order to be able to trigger custom logic before or after changes to the DB. 

one use case is to create an audit trail table in the DB.
I  want to have an Audit Trail table in the DB to track all changes made on certain table. Something like:
TableName
RecordId
UserName
Data
ChangeType (Update/ Insert/ Delete)
OldValues (JSON)
NewValues (JSON)

I want to be able to register to an event just after (or before ) changes are committed to the DB in order to prepare the data based on the changes that were done and save it to the Audit Trail table.

## Solution
Add life cycle events before and after changes are committed to the DB.
In these event the developer should have simple API to get a list of all changes that were committed to about to be committed. 
The developer should have an option to add new records during this event, or make changes to the data before it is committed. 
In cases when using the "before" events the developer should be able to make the changes in context of the original transaction.

## Alternatives
Another option, on top of the previous one, is to have a built in Audit Trail mechanism where the developer will be able per table to configure whether it should participate or not in the audit trail process, and if it does Prisma should add Audit Trail records automatically to the Aduit Trail table.

**Comment by divyenduz** [comment]
@yuval-hazaz Thank you for raising this. This can also be done with SQL level triggers right? Have you considered that as a use case for solving your needs?

**Comment by yuval-hazaz** [comment]
@divyenduz  
Yes. this can be done using triggers on the DB but it has several disadvantages. 
1. It requires adding a trigger for each of the tables in the DB.
2. It requires "knowing" your DB and adding logic at the DB side. It impacts the biggest advantage of prisma which is DB agnostic.  
2. When using triggers it can impact performance as the audit trail records will be committed as part of the transaction. The new feature should allow adding the records after the transaction is committed to improve performance.

**Comment by mrsimply** [comment]
@divyenduz 

Another disadvantage of DB triggers is, that the DB does not have a context.
I want to create an Audit Trail too this way - Using triggers I can not log the "current user" logged in on the application side.

I would really appreciate @yuval-hazaz described approach - I already tried a combination of Prisma Middleware and DB triggers, but non of it comes handy. E.g. Middleware lags to catch events for models during nested writes.

**Comment by janpio** [comment]
> E.g. Middleware lags to catch events for models during nested writes.

If this is a limitation, it would be great if you could open a separate issue about this. Thanks.

**Comment by faoiseamh** [comment]
+1 for the life cycle event hook approach.

I think the best solution to this and other similar problems is to provide hooks / callbacks into the transaction lifecycle so that we can interact with the transaction / query lifecycle at the application level where we have access to more context than the database level. This is both more powerful and more portable than database-level triggers as it doesn't depend on the specific database engine and has a greater available context including the other application functionality.

**Comment by lucasmarinzeck** [comment]
Being able to use a built-in solution for audit trial configurable by table would be great

**Comment by mayerlench** [comment]
Iv ran into this issue as well when trying to implement some kind of audit trail

**So far here is what i was able to accomplish with DB level triggers**

- TableName
- RecordId
- ChangeType (Update/ Insert/ Delete)
- OldValues (postgres hstore)
- NewValues (postgres hstore)

The main issue like @mrsimply explains is, theres no context in the database. We can know which user the query was triggered by.

I think @yuval-hazaz had a good idea about having a feature built-in for audits. That would be really cool

**Comment by ACPK** [comment]
@mrsimply @mayerlench 



> Iv ran into this issue as well when trying to implement some kind of audit trail
> 
> **So far here is what i was able to accomplish with DB level triggers**
> 
> * TableName
> * RecordId
> * ChangeType (Update/ Insert/ Delete)
> * OldValues (postgres hstore)
> * NewValues (postgres hstore)
> 
> The main issue like @mrsimply explains is, theres no context in the database. We can know which user the query was triggered by.

I hate to add another field to the database, but could you solve this today by adding an "author" column to the database?

The "author" column can be (1) the userId from the application or (2) hard-coded with the data source when doing a "CSV" import.

**Comment by Saladinha** [comment]
I would like to see the same thing, some prisma builtin audit trail, considering maybe having automatic generated audit tables to avoid the need of json for old values and new values it could be e.g

post_audit
____________
id (post_audit PK)
postId (post FK)
_prev_title 
_title
_prev_post_content
_post_content
_prev_published
_published
operationType(like @yuval-hazaz  said Update, Insert, Delete)
changedBy (maybe user FK depends on project)
changedAt (datetime)

**Comment by joesaunderson** [comment]
Is there any update on this feature request? I have tried implementing it via `Middleware` but this only shows the action and the model - as mentioned above there needs to be a method for gaining context of a model to record the before and after state.

**Comment by pedroeckel** [comment]
I would also like to know if we have had any updates on this feature request.

**Comment by agustif** [comment]
https://github.com/unlight/prisma-query-log
https://access.crunchydata.com/documentation/pgaudit/1.2.0/

What I found that might be useful

**Comment by muhammadshaban** [comment]
Can we have events like this
https://laravel.com/docs/9.x/eloquent#events
It will be very helpful for audit

**Comment by gilbertronaldo** [comment]
any update about this feature?  
maybe the feature is like [https://github.com/owen-it/laravel-auditing](https://github.com/owen-it/laravel-auditing)

**Comment by Blanen** [comment]
Most DBs have audit logging features or extensions. These can work much more efficiently.
And to store the audit log of a database in the database itself might be problematic anyway.

**Comment by faoiseamh** [comment]
> Most DBs have audit logging features or extensions. These can work much more efficiently. And to store the audit log of a database in the database itself might be problematic anyway.

Agreed that there are definitely viable database-level audit log options if all we want is a log of database operations, but for my uses-cases, interacting with changes in a subset of database records over time is a core element of the application and requires combining the low level CRUD operations with application-level context from the session responsible for the changes. That's the reason database-level logging is not an ideal solution – there's not an elegant way to integrate application-level (session, request) context into the audit log entries.

Of course we could explicitly insert these logs alongside each operation we want controlled in this fashion, but that requires remembering to do this explicit logging at every possible place one of our "controlled" entities was being created, updated, or deleted and for something as critical as this is for us, that's unacceptable. To ensure absolutely nothing is missed, we need to keep our logging as close to the database as possible while stlil having access to application context.

That's why a middleware solution incorporated into our database abstraction layer is the ideal solution, and a solution I've implemented very effectively in other frameworks like Rails' ActiveRecord, where this audit logging process ("objection versioning") is implemented one time in a mixin that is used to compose every model we want monitored like this, hooking into the model's lifecycle events and adding a final layer of log object creation.

I understand Prisma is a different sort of (non) ORM, so this type of solution doesn't fit with a framework like this. But the middleware offered by Prisma is _so close_ to being able to elegantly handle this use case that it seems like a worthwhile use-case to consider. I'm sure there are a variety of other scenarios where access to application / request context from middleware would prove valuable.

**Comment by HendrikJan** [comment]
> 

An "author" column could work when using DB triggers, except for delete actions.
Hooks in Prisma (like in Sequelize) would make the problem much easier to solve.

**Comment by aserranom** [comment]
We finally opted for adding an `updatedBy` field on every single table and creating a middleware that recursively sets the `updatedBy` to every update call based on who invoked it. Our data team then generates logs using a AWS' DMS task and some post processing.

Far from perfect but it somewhat suits our use case for now.

> Prisma is so close to being able to elegantly handle this use case
I agree with @faoiseamh, so close!

@divyenduz is there any way we could help into making this a reality?

**Comment by faoiseamh** [comment]
We're using client extensions fairly effectively to achieve this entirely in prisma currently, you might try that approach @maZahaca

**Comment by maZahaca** [comment]
@faoiseamh Thanks for pointing me to the right direction, are you able to share a code snippet? I'm quite new to Prisma.

**Comment by NicholasFields** [comment]
Echoing the notion above that hooks could make this a lot simpler [and more flexible] than SQL triggers.

In Objection, for example, you can fire off arbitrary actions when a model underoes a given procedure [insert, update, etc.].

Looking into other (SQL) audit-logging methodolies both inside and outside of Prisma. 

Including but not limited to:
- AWS DMS (mentioned above)
- AWS Event Bus 
- Pure custom logic at the application layer

Anyone knows some good stuff, please LMK.

**Comment by kairauer** [comment]
@maZahaca came back to this issue as well and checked how client extensions could solve this. [https://github.com/prisma/prisma-client-extensions this migh](https://github.com/prisma/prisma-client-extensions/tree/main/audit-log-context) this from their official examples might help

**Comment by codechikbhoka** [comment]
This seems to be an effective solution: 
https://debezium.io/blog/2019/10/01/audit-logs-with-change-data-capture-and-stream-processing/

**Comment by shadoath** [comment]
@faoiseamh or @aserranom can either of you provide links to examples of what you did?

**Comment by arjunlol** [comment]
If going the CDC route as @codechikbhoka mentioned, it can still get tricky to pass the context (userID, API endpoint, etc..) down to the infrastructure level with Prisma. What I'd recommend is injecting this application metadata via sql comments with the raw query. This pattern is typically used in [database tracing](https://cloud.google.com/blog/topics/developers-practitioners/introducing-sqlcommenter-open-source-orm-auto-instrumentation-library) and would need to use [Prisma's new driver adapter preview feature](https://www.prisma.io/docs/orm/overview/databases/database-drivers#driver-adapters). You can check out the source code of how we did this here: https://github.com/BemiHQ/prisma

**Comment by ravivarma4003** [comment]
Any update on this feature request ?

**Comment by ravivarma4003** [comment]
@janpio : Any update on Above request? It will be better if Prisma Documentation Explicitly covers this use case with extension approach if it is possible with https://www.prisma.io/docs/orm/prisma-client/client-extensions/query#extend-prisma-client-query-operations.

**Comment by ravivarma4003** [comment]
@faoiseamh Can you please explain or share a basic code snippet on how did you handled audits using prisma client extensions . Will be happy to see update audits

**Comment by exAspArk** [comment]
Check out https://github.com/BemiHQ/bemi-prisma, which we built to implement an automatic audit trail.

It uses a Prisma extension and the [`pg`](https://github.com/brianc/node-postgres) driver:

![image](https://github.com/prisma/prisma/assets/577441/3367365d-8c84-4851-9288-3549262bb05e)

**Comment by Erry91** [comment]
Just so that is clearer to me, the bemi-prisma integration requires a subscription right?

**Comment by exAspArk** [comment]
@Erry91 yes, Bemi requires a subscription after a 14-day trial. We're also building on a self-hosted solution that runs all processes for ingesting data changes using database replication logs to keep a 100% reliable and accurate audit trail.

**Comment by Dkyc** [comment]
@exAspArk What happens when you do `npx prisma generate` and `npx prisma db push`

**Comment by 0xdevalias** [comment]
> @maZahaca came back to this issue as well and checked how client extensions could solve this. [https://github.com/prisma/prisma-client-extensions](https://github.com/prisma/prisma-client-extensions/tree/main/audit-log-context?rgh-link-date=2023-05-03T20%3A33%3A41Z) this from their official examples might help

Since the displayed version of this link hides the actual direct URL of the example it points to, I just wanted to make it more visible for others searching for it:

- https://github.com/prisma/prisma-client-extensions/tree/main/audit-log-context
  - > Prisma Client Extension - Audit Log Context
    > This example shows how to use a Prisma Client extension to provide the current application user's ID as context to an audit log trigger in Postgres. User IDs are included in an audit trail tracking every change to rows on certain tables.

**Comment by nkolly** [comment]
Sheesh open for five years, is this a priority at all or should I look for other options at this point?

**Comment by agustif** [comment]
> Sheesh open for five years, is this a priority at all or should I look for other options at this point?

Seems this is certainly not a focus of Prisma as a framework.

I've been happy to find -alternatives- that play nicely with it where I want to extend and it doesn't allow due to it's constraints, namely just found out about bemi https://bemi.io/ https://medium.com/@arjunlall/prisma-audit-trail-guide-for-postgres-5b09aaa9f75a and this seems like the -better- way to do it lacking official prisma support, pgAudit is also pretty robust I think.

I also quite liked finding out about zenstack to add RBAC and ABAC on a similar DSL than prisma's schema, and it seems a project named AtlasGo that offers RLS for pg/prisma too https://atlasgo.io/guides/orms/prisma/row-level-security

all in all postgres is a great engine and as long as all the pieces you're plugging to it make sense together you should be fine (easier said than done, I know)

**Comment by moinulmoin** [comment]
prisma really sucks at this, (not solving issues for a long time) lol

**Comment by happy1836** [comment]
Model should be allowed to use "extend"

**Comment by dvecy** [comment]
For anyone needing this today - I built a lightweight TypeScript SDK called trailkit that handles exactly this: actor/resource/action events, AsyncLocalStorage context propagation (no manual threading), SHA-256 hash chaining for tamper detection, and adapters for SQLite/PostgreSQL. Zero infrastructure required. https://github.com/sadatnazarli/trailkit
